File: src/day_11/part1.erl
-module(part1).
-export([main/0, count_paths/1, parse_input/1]).

-define(START_NODE, "you").
-define(END_NODE, "out").
-define(INPUT_FILENAME, "input.txt").


main() ->
    case file:read_file(?INPUT_FILENAME) of
        {ok, BinaryInput} ->
            Input = unicode:characters_to_list(BinaryInput),
            Graph = parse_input(Input),
            Result = count_paths(Graph),
            io:format("~p~n", [Result]);
        {error, Reason} ->
            io:format("Error reading file ~s: ~p~n", [?INPUT_FILENAME, Reason]),
            halt(1)
    end.

parse_input(Input) ->
    Lines = string:split(Input, "\n", all),
    lists:foldl(
        fun(Line, Acc) ->
            case parse_line(Line) of
                {ok, Device, Outputs} ->
                    Acc#{Device => Outputs};
                {error, _} ->
                    Acc
            end
        end,
        #{},
        Lines
    ).

parse_line(Line) ->
    case string:split(string:trim(Line), ":", leading) of
        [NameStr, OutputsStr] ->
            Name = string:trim(NameStr),
            Outputs = [string:trim(O) || O <- string:split(string:trim(OutputsStr), " ", all), O /= ""],
            {ok, Name, Outputs};
        _ ->
            {error, "Invalid line format"}
    end.

count_paths(Graph) ->
    {_, Count} = count_paths_recursive(Graph, #{}, ?START_NODE),
    Count.

count_paths_recursive(_Graph, Memo, Device) when Device =:= ?END_NODE ->
    {maps:put(Device, 1, Memo), 1};

count_paths_recursive(Graph, Memo, Device) ->
    case maps:find(Device, Memo) of
        {ok, Count} ->
            {Memo, Count};
        error ->
            Outputs = maps:get(Device, Graph, []),
            
            {FinalMemo, TotalPaths} = lists:foldl(
                fun(OutputDevice, {CurrentMemo, Acc}) ->
                    {NewMemo, Paths} = count_paths_recursive(Graph, CurrentMemo, OutputDevice),
                    {NewMemo, Acc + Paths}
                end,
                {Memo, 0},
                Outputs
            ),

            {maps:put(Device, TotalPaths, FinalMemo), TotalPaths}
    end.

File: src/day_11/part1.hs
import Data.Map (Map)
import qualified Data.Map as Map
import Data.List (foldl')
import System.IO (getContents)

type Device = String
type Graph = Map Device [Device]

parseInput :: String -> Graph
parseInput input =
    Map.fromList $ map parseLine (lines input)
  where
    parseLine :: String -> (Device, [Device])
    parseLine line =
        case words $ filter (/= ':') line of
            (name : outputs) -> (name, outputs)
            _                -> error $ "Invalid line format: " ++ line

countPaths :: Graph -> Integer
countPaths graph =
    let
        memo :: Map Device Integer
        memo = Map.empty

        go :: Map Device Integer -> Device -> (Map Device Integer, Integer)
        go currentMemo device
            | Just count <- Map.lookup device currentMemo = (currentMemo, count)
            | device == "out" = (Map.insert device 1 currentMemo, 1)
            | otherwise =
                case Map.lookup device graph of
                    Nothing -> (Map.insert device 0 currentMemo, 0)
                    Just outputs ->
                        let
                            (finalMemo, totalPaths) = foldl'
                                (\(m, acc) outputDevice ->
                                    let
                                        (newMemo, paths) = go m outputDevice
                                    in
                                        (newMemo, acc + paths)
                                )
                                (currentMemo, 0)
                                outputs
                        in
                            (Map.insert device totalPaths finalMemo, totalPaths)
    in
        let (_, finalCount) = go memo "you"
        in finalCount

main :: IO ()
main = do
    input <- getContents
    let graph = parseInput input
        result = countPaths graph
    print result